pageserver/src/tenant/timeline.rs

```diff
@@ -2223,6 +2228,38 @@ impl Timeline {
             layer_paths.pop().unwrap();
         }
 
+        drop(all_keys_iter); // So that deltas_to_compact is no longer borrowed
+
+        Ok(CompactLevel0Phase1Result {
+            new_layers,
+            deltas_to_compact,
+        })
+    }
+
+    ///
+    /// Collect a bunch of Level 0 layer files, and compact and reshuffle them as
+    /// as Level 1 files.
+    ///
+    async fn compact_level0(&self, target_file_size: u64) -> anyhow::Result<()> {
+        let CompactLevel0Phase1Result {
+            new_layers,
+            deltas_to_compact,
+        } = self.compact_level0_phase1(target_file_size).await?;
+
+        // Before deleting any layers, we need to wait for their upload ops to finish.
+        // See storage_sync module level comment on consistency.
+        // Do it here because:
+        // 1. We don't want to hold self.layers.write() while waiting
+        // 2. We know that the new layers aren't going to be deleted, so, we can
+        //    avoid waiting for them.
```

Comment:
Hmm. There are two operations that delete files: compaction and GC. As long as we have something that prevents GC from running at the same time, it's true.

Reply:
The comment implies that they're not going to be removed *because* they're new though. Not because we're preventing GC from running at the same time. I'm not sure about that.

Reply:
> Hmm. There are two operations that delete files: compaction and GC. As long as we have something that prevents GC from running at the same time, it's true.

This holds due to `layer_removal_cs`.

> The comment implies that they're not going to be removed because they're new though. Not because we're preventing GC from running at the same time. I'm not sure about that.

Me neither. The assert never triggered, though.

**Pro:** It's a clear latency win over doing wait_completion() after scheduling the level 1 layers for upload. Compaction latency, that is.
**Con:** Then again, does compaction latency matter at all? If we're producing compacted layers faster than upload rate, we will fall behind in the next compaction iteration.

Reply:
Looking up the call stack:
1. We measure `Timeline::compact_level0()` latency, so apparently it matters to someone (`compact_time_histo`).
1. Timeline::compact() is invoked sequentially, in a loop.
1.  `compaction_loop` itself does a fixed amount of `sleep`, i.e., it doesn't subtract the time it took to do the compaction.

Not sure what to make of that.

Reply:
It's not latency-critical as such, but if compaction is too slow, it might not keep up with the speed at which new WAL is ingested.

The code looks good, let's just change the comment to "Do it here because we don't want to hold self.layers.write() while waiting", leaving out the reason (2).